src/TipToiYaml.hs

```diff
@@ -164,11 +170,37 @@ scriptCodes codes codeMap
             Nothing -> Left s 
             Just n -> Right (n::Word16)
 
+-- The following logic (for objectCodes) tries to use different object codes 
+-- for different projects, as far as possible. This makes the detection of not 
+-- having activated a book/product more robust.
+
+-- We could theoretically set: 
+--    objectCodeOffsetMax = lastObjectCode - firstObjectCode.
+-- This would assign perfectly usable object codes, and would minimize the 
+-- probability of object code collisions between products, but sometimes 
+-- object codes would wrap around from 16383 to 1000 even for small projects 
+-- which may be undesirable. We arbitrarily do not use the last 199 possible 
+-- offsets to avoid a wrap around in object codes for projects with <= 200 
+-- object codes. This does not impose any limit on the number of object codes 
+-- per project. Every project can always use all 15384 object codes.
+    objectCodeOffsetMax = lastObjectCode - firstObjectCode - 199
```

Comment:
Ouch! No, I must admit I do not know much about the pen and the GME format and 
was blindly assuming the GME file contains a dictionary of arbitrary opaque OIDs to scripts.

If the OIDs need to be chosen from one consecutive range then we should never wrap around and
I completely agree that we then must assume a worst case number of OIDs at the end of the range.
What about 2000? There is relatively little benefit in minimizing this compared to the 'cost'
of potentially generating broken GME files.

I will change the 199 to 1999 and remove the wrap around logic and will submit a new patch.

Thanks for responding to this.

Cheers,
Johannes

Am 11.10.2015 um 12:15 schrieb Joachim Breitner <notifications@github.com>:

> In src/TipToiYaml.hs:
> 
> > @@ -164,11 +170,37 @@ scriptCodes codes codeMap
> >              Nothing -> Left s 
> >              Just n -> Right (n::Word16)
> >  
> > +-- The following logic (for objectCodes) tries to use different object codes 
> > +-- for different projects, as far as possible. This makes the detection of not 
> > +-- having activated a book/product more robust.
> > +
> > +-- We could theoretically set: 
> > +--    objectCodeOffsetMax = lastObjectCode - firstObjectCode.
> > +-- This would assign perfectly usable object codes, and would minimize the 
> > +-- probability of object code collisions between products, but sometimes 
> > +-- object codes would wrap around from 16383 to 1000 even for small projects 
> > +-- which may be undesirable. We arbitrarily do not use the last 199 possible 
> > +-- offsets to avoid a wrap around in object codes for projects with <= 200 
> > +-- object codes. This does not impose any limit on the number of object codes 
> > +-- per project. Every project can always use all 15384 object codes.
> > +    objectCodeOffsetMax = lastObjectCode - firstObjectCode - 199
> 
> Shouldn�t this be much larger? A wrap-around means that the gme file becomes very large (one word per oid code in range). Have you made sure that this is not a problem for the pen?
> 
> �
> Reply to this email directly or view it on GitHub.
> 

-- 
Johannes.Overmann@joov.de (http://joov.de)
Fichtenstr. 5, 79194 Gundelfingen, +49 761 556 3222